core/src/main/java/org/eclipse/hono/util/EventDemultiplexer.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.util;
+
+import org.apache.qpid.proton.amqp.messaging.Data;
+import org.apache.qpid.proton.amqp.messaging.Section;
+import org.apache.qpid.proton.message.Message;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * A helper for handling different types of events with explicit callbacks.
+ * Currently it demultiplexes callbacks for generic events and notifications (being well defined events).
+ */
+public final class EventDemultiplexer {
+    /**
+     * Prevent construction.
+     */
+    private EventDemultiplexer() {
+    }
+
+
+    /**
+     * Construct a event consumer that delegates different kinds of events to specific handlers.
+     * @param eventConsumer Consumer that handles generic events.
+     * @param notificationConsumer Consumer that handles events that were detected to be a notification. Maybe {@code null}.
+     * @return The event consumer that invokes the specific handlers depending on the event type.
+     */
+    public static final Consumer<Message> createEventHandler(final Consumer<Message> eventConsumer,
+                                                             final Consumer<Message> notificationConsumer
```

Comment:
Wouldn't it be an idea to put in a Map here and map from Subject to consumer?

Reply:
@ctron: thanks for the idea!
But I am afraid I do not understand what you mean by it - currently this class just demultiplexes based on the message, and the consumers must be provided by the user of this class (as lambdas, as method references, etc.).

Where could a `Map` improve things?

Reply:
So currently the demuxing is done based on the result of "getContentType", which is a String IIRC.

So instead of using a single `Consumer<Message>` as the second argument, we could use a `Map<String,Consumer<Message>>` and use the result of `getContentType` as key. Something like:

~~~
String type = message.getContentType();
Consumer<Message> specialConsumer = consumerMap.get(type);
if ( specialConsumer != null ) {
  specialConsumer.accept ( message );
} else {
  defaultConsumer.accept ( message );
~~~

This would allow to use the demuxer with additional content types without the need to add new arguments in the "createEventHandler" all the time.

Reply:
@ctron With the current specification, there is no "device command readiness notification" based on a content-type anymore.
So the good idea currently does not make so much sense anymore.
See PR #596 discussions for details, and thanks again for the comment!